### PR TITLE
xmr: hardfork 10 upgrade

### PR DESCRIPTION
**Pull request by ph4r05**
Monero hardfork 10 upgrades:

- Code now supports only hardforks 9 and 10 (Borromean range sig removed, Full MLSAG signature removed).
- Deterministic output commitment masks, based on amount_key
- Bulletproof v2 serialization, EcdhInfo serialized as 8 B amount, XOR encrypted by a specific key derived from amount key
- Signing: pseudo_out recomputation on inputs, sign step. Due to the deterministic output masks, balancing has to be done on input commitments. As HF8 support is discontinued, pseudo_out are not hashed in the signature as it has been moved to prunnable section in HF9.
- Dummy encrypted payment ID (if applicable) for better transaction uniformity (implemented in Monero wallet already)

Requires https://github.com/trezor/trezor-common/pull/280 to be merged for tests to work. Test are ready but I need to version-bump python-trezor and monero-agent.

## Deterministic masks
Since HF10, the output masks are deterministic, based on the amount_key. The decision is the deterministic masks will be used also in HF9 if client supports this feature, i.e., `client_version=1`. 

Currently, `client_version=0` clients do not support deterministic output masks. The main reason is the mask balancing (i.e., sum output masks equals sum of input masks) was performed on the output masks side. The `client_version=0` clients do not support the balancing on input side (there is no way for the client to update the masks after balancing).

However, masks are generated deterministically also in `client_version=0`, in the step 5 (not using amount key). This helps to lower memory footprint and makes code more unified.

Once all clients are updated, the old-way of generating masks will be removed from the step 5.

## Protocol change

Due to the deterministic output masks the protocol was changed a bit. The masks are computed from the amount_keys which are computed in the step 6 (set output). 

This affects the BP offloading logic. Masks need to be sent to the host so it can compute the BP. Thus the offloading can be done only after all outputs in the current batch have been generated.

In the BP offloading scenario the protocol goes as following:
- `MoneroTransactionSetOutputRequest` per output (as before). The response `MoneroTransactionSetOutputAck` contains corresponding `mask` in the `rsig_data` field.
- A new `MoneroTransactionSetOutputRequest(is_offloaded_bp=1)` message is sent by the host which contains computed BP in the `rsig_data` field after all masks from the current batch has been computed by the Trezor.

The masks balancing can be performed after all output masks were computed, thus no earlier than in step 7. The balancing is thus performed on the input masks in the step 9 (signing).

The last input mask in the sequence is modified so the sum of input and output masks equals. The host need to know about this change so it can update the transaction structures. The `MoneroTransactionSignInputAck` message now contains `pseudo_out` field which carries the updated output commitment value.

### CODE REVIEW

**Comment by ph4r05** [comment]
PR description updated

**Comment by ph4r05** [comment]
trezor-common version bump

**Comment by ph4r05** [comment]
From some reason, `test_apps.common.coins.py` and `test_trezor.config.py` tests are failing now. Idea why? 😕

**Comment by ph4r05** [comment]
Hmm my monero tests are passing locally but failing on Travis. I suspect it is because of outdated python-trezor.

**Comment by tsusanka** [comment]
(I'm looking into it)

**Comment by tsusanka** [comment]
It should be fixed. I wanted to do a rebase and push, but it seems I can't push into this PR.

**Comment by ph4r05** [comment]
Great! The tests are now passing. The tests perform TX signature for `client_version=0` HF9. Additionally, the also `client_version=1` HF9 and HF10 are tested.

**Comment by tsusanka** [comment]
I will do those myself. Soon

File: src/apps/monero/signing/step_01_init_transaction.py [hunk 1]
```diff
@@ -293,18 +294,24 @@ def _process_payment_id(state: State, tsx_data: MoneroTransactionData):
...
-    # encrypted payment id
-    if len(tsx_data.payment_id) == 8:
+    if not tsx_data.payment_id or len(tsx_data.payment_id) == 8:
         view_key_pub_enc = _get_key_for_payment_id_encryption(
-            tsx_data.outputs, state.change_address()
+            tsx_data, state.change_address(), state.client_version > 0
         )
 
+    if not tsx_data.payment_id:
+        return
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply src/apps/monero/signing/step_01_init_transaction.py]: I made this maybe in a bit confusing way. The `_get_key_for_payment_id_encryption()` is the only place which checks if short payment ID is applicable (checks if TX has 2 outputs, one change). 

This function can add dummy payment ID if there is none provided by the host so the check on payment_id None is moved after the call to `_get_key_for_payment_id_encryption`.

[the remark this answers is not in the document]
Comment (tsusanka) [reply src/apps/monero/signing/step_01_init_transaction.py]: Ah, ok. It is a bit misleading, but I think we can live with that.

**Comment by tsusanka** [comment]
nit-picks implemented in https://github.com/trezor/trezor-core/commit/2b048ce25cb96f9fed64992e2a154c6b930d1761

**Pull request merged**